xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/async/MapBasedLinkedBlockingQueue.java

```diff
@@ -0,0 +1,364 @@
+            Pair<Long, T> castedPair = (Pair) pair;
+            this.internalMap.remove(castedPair.getKey());
+        }
+    }
+
+    /**
+     * Clear the internal map and reset the highest key to 0.
+     */
+    private void clearInternalMap()
+    {
+        this.internalMap.clear();
+        this.highestKey.set(0);
+    }
+
+    @Override
+    public void put(T t) throws InterruptedException
+    {
+        Long key = this.internalPut(t);
+        this.internalQueue.put(Pair.of(key, t));
+    }
+
+    @Override
+    public boolean offer(T t, long l, TimeUnit timeUnit) throws InterruptedException
+    {
+        Long key = this.internalPut(t);
+        return this.internalQueue.offer(Pair.of(key, t), l, timeUnit);
+    }
+
+    @Override
+    public boolean offer(T t)
+    {
+        Long key = this.internalPut(t);
+        return this.internalQueue.offer(Pair.of(key, t));
+    }
+
+    @Override
+    public T take() throws InterruptedException
+    {
+        Pair<Long, T> result = this.internalQueue.take();
+        this.internalRemove(result);
+        return result.getValue();
+    }
+
+    @Override
+    public T poll(long l, TimeUnit timeUnit) throws InterruptedException
+    {
+        Pair<Long, T> result = this.internalQueue.poll(l, timeUnit);
+        this.internalRemove(result);
+        return result.getValue();
+    }
+
+    @Override
+    public int remainingCapacity()
+    {
+        return this.internalQueue.remainingCapacity();
+    }
+
+    @Override
+    public T poll()
+    {
+        Pair<Long, T> result = this.internalQueue.poll();
+        this.internalRemove(result);
+        return result.getValue();
+    }
+
+    @Override
+    public T element()
+    {
+        return this.internalQueue.element().getValue();
+    }
+
+    @Override
+    public T peek()
+    {
+        Pair<Long, T> result = this.internalQueue.peek();
+        return result.getValue();
+    }
+
+    @Override
+    public boolean remove(Object o)
+    {
+        boolean result = this.internalQueue.remove(o);
+        this.internalRemove(o);
+        return result;
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> collection)
+    {
+        for (Object o : collection) {
+            if (!this.contains(o)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public int size()
+    {
+        return this.internalQueue.size();
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return this.internalQueue.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o)
+    {
+        // We check on the map since most likely the given object is not a Pair but a T.
+        return this.internalMap.containsValue(o);
+    }
+
+    @Override
+    public Iterator<T> iterator()
+    {
+        return this.internalMap.entrySet().stream().sorted(new DefaultComparator()).map(Map.Entry::getValue).iterator();
```

Comment:
Why not `internalQueue.iterator()` ?

Reply:
Because the internal queue is of type `Pair<Long, T>` where I need an Iterator of T.

Reply:
You could map it. It would still be much faster than manipulating serialized elements.

By the way to be complete you should also wrap the Iterator because of Iterator#remove().